feat: add constraint stream profiling

Christopher-Chianelli opened the pull request:
Profiling constraints is notorishing difficult, since each component of a constraint are converted to nodes, some of which are shared. As such, a JVM method profile is basically unreadable and does not represent how much time is actually spent for each constraint.

To aid in profiling, an optional constraintStreamProfilingMode configuration was added. If set to a value other than NONE, it wraps each tuple lifecycle node inside a ProfilingTupleLifecycle, which will measure how long each lifecycle executes. The ProfilingTupleLifecycle find out what constraint is responsible for creating that lifecycle by getting snapshot of the stack traces from its constraint stream's creator (when a constraint stream is shared, their stack traces are merged into the same set).

At the end of solving, a profiling summary is then produced in the INFO log. The details differ depending on the profiling mode:

- In the BY_METHOD profiling mode, (className, methodName) is used as the key

- In the BY_LINE profiling mode, (className, methodName, lineNumber) is used as the key.

The methods/lines are printed in descending order of time percentage spent. The sum of time percentage spent may be over 100%, since methods/lines can share time spent with other methods/lines.

timefold.solver.constraint-stream-profiling-mode was added as a property to Quarkus and Spring Boot to configure profiling (defaults to NONE).

triceo commented on benchmark/src/main/resources/benchmark.xsd:
```diff
@@ -458,6 +458,9 @@
           <xs:element minOccurs="0" name="constraintStreamAutomaticNodeSharing" type="xs:boolean"/>
                               
           
+          <xs:element minOccurs="0" name="constraintStreamProfilingMode" type="tns:constraintProfilingMode"/>
```
In the interest of not asking the user to make niche choices, I'd remove this. See below.

triceo replied:
I think we need a better name than "profiling".
It's confusing the terms - there is no profiler here, there is no profile, IMO this name does more harm than good.

triceo commented on core/src/main/java/ai/timefold/solver/core/config/score/director/ConstraintProfilingMode.java:
```diff
@@ -0,0 +1,22 @@
+ * Controls how constraints are profiled.
+ * Enabling profiling have a minor performance impact.
+ */
+public enum ConstraintProfilingMode {
```
IMO this is needless complexity. Do not profile by method or line - profile _by constraint_. That is what people actually want to see; the list of constraints and how much time they take to run.

Also, the names of constraints typically do not change, while method names and lines do while code is refactored over time. This should IMO not be information we depend on, or display.

triceo reviewed (commented):
I like this in principle! When it comes to the implementation, I have concerns. See specifically the comment regarding profiling mode.

I also see this hooks itself into tuple lifecycle. But I don't see it tracking propagation overhead anywhere. Arguably it should.

The logging aspect is OK, but I am missing any methods to consume this information at runtime. How about Micrometer metrics? Or custom JFR events? Thinking out loud here: if it triggered an event (micrometer / JFR) every time it finished a single measurement, these could be aggregated at the end as well as queried in-flight. More often than not, we will be run in the platform - people will want a way of monitoring their solvers.

I question the name "profiling"; this doesn't use a profiler, and doesn't work in a way in which profilers work. The name should be changed in order to not invite confusing comparisons with actual profiling.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/bavet/common/LifecycleKind.java:
```diff
@@ -0,0 +1,15 @@
+package ai.timefold.solver.core.impl.bavet.common;
+
+public enum LifecycleKind {
```
This needs documentation. What's the point? Are all handed the same? Are all from constraint streams? Wouldn't it be better to look forward through the lifecycle to the actual node? (Currently not possible, but _could be_.)

This is brittle, and effectively duplicates node logic. Also, the fact that ifExists is represented by join (and therefore not measurable) is unfortunate. IMO needs to be addressed.

triceo reviewed (commented):
The output could use being shorter and clearer.
The implementation around `LifecycleKind` needs rethinking.

Christopher-Chianelli commented on core/src/main/java/ai/timefold/solver/core/impl/bavet/common/tuple/LeftTupleLifecycleImpl.java:
```diff
@@ -10,6 +12,11 @@ record LeftTupleLifecycleImpl<Tuple_ extends Tuple>(LeftTupleLifecycle<Tuple_> l
 
+    @Override
+    public LifecycleKind lifecycleKind() {
+        return LifecycleKind.JOIN;
```
This is for us; users would rather have the location than the node type. Although `concat`, `ifExists` and `join` are different operations, they all fundamentally depends on a "joining" conceptually. Would you rather it be called `LEFT`/`RIGHT`?

triceo commented:
> When it is alone, it means the work done exclusively by a given constraint (where removing the constraint guarantees removing that work).
Arguably if this case used different words, I wouldn't have to ask the question.
Different cases require different descriptions.

Acknowledged, but it is the raw captured data. The issue is if we try to humanize the number, then we need to ask questions such as do we convert it to minutes above a certain duration? Then hours? And then there the potential of losing accuracy and making comparisons between constraints harder (as you need to convert them to a common unit).
You already picked a common unit, just the worst one. :-) I see how this could possibly be hours, days. Or nanos. Practically, it won't be either. It will not be much less (who stops the solver after a millisecond?), and it will not be much more (who needs hours to establish these numbers?).

IMO if we use seconds as a baseline, that's a reasonable estimate. Saying "5:25:00" (hms) also isn't too hard, if we decide we need to. The "problem" comes when the time is much smaller than a second - thankfully, it will not be.

> Useful information for who? This information also shows up in the location list.

You probably noticed that I don't like the location list.

> This is the only text that is actually actionable by the user; it tells them what is actually hot in their code, and they don't need to do a string/constant lookup.
Since it doesn't cross-link to code, you're copy-pasting anyway. IMO constraint names work much better; those will always be the same. With line numbers, you're making many assumptions; that the code over there is the same as the one you're looking at, that the code was compiled to include those line numbers, that no crazy tooling modified that code. (I love Lombok, which always points me to places that don't really exist.)

I know that this is one way how you can report node sharing. IMO the downsides outweigh that; this will not be reliable, and only leads to questions.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/bavet/common/ConstraintProfile.java:
```diff
@@ -0,0 +1,134 @@
+    private String formatDuration(long nanos) {
+        var NANOS_IN_SECOND = 1_000_000_000L;
+        var seconds = (double) nanos / NANOS_IN_SECOND;
+        return "%.2f seconds".formatted(seconds);
```
I'd prefer `Duration.toSeconds()` and `toMillisPart()`.

triceo reviewed (commented):
Some more comments, but we're getting there. The output is much more readable now!

Needs documentation. In there, it specifically needs to explain what it measures and how (wall clock / CPU time / ...), what is the expected impact on CS performance, and how we expect people to use this. (Long runs? Production?) It should also break down an example output, describe its components and how to understand the information given. Come to think of, it should even mention _where_ the information will be printed and how to make sure that it is. (Logging config.)

I'm still not wild about the world "profiling". Is it profiling if there's not a profiler? No Java agents, nothing resembling the usual profiling tools for the JVM? Could lead to confusion of terms. What are our other options for naming?

triceo reviewed (approved):
LGTM once documented.

TomCools commented on docs/src/modules/ROOT/pages/constraints-and-score/performance.adoc:
```diff
@@ -315,6 +315,145 @@ Whatever you do, benchmark on a large and diverse set of inputs.
+
+* A breakdown for each constraint (in descending order of time spent):
+
+** The time spent (relative percentage of time spent by constraint and total seconds), by each constraint stream building block (such as `join` or `filter`).
```
```suggestion
** Time spent by each constraint stream building block (e.g., join, filter), shown as % and total seconds.
```

The double "time spent" mention makes it harder to read.

TomCools commented on docs/src/modules/ROOT/pages/constraints-and-score/performance.adoc:
```diff
@@ -315,6 +315,145 @@ Whatever you do, benchmark on a large and diverse set of inputs.
++
+NOTE: This info is only relevant to the Timefold Solver team and can be safely ignored.
+
+*** The time spent (relative percentage of time spent by building block and total seconds) of each defined node and the location(s) the nodes were defined at.
```
```suggestion
*** Time spent by each constraint stream building block, shown as % and total seconds of each each defined node and the location(s) the nodes were defined at.
```

TomCools commented on docs/src/modules/ROOT/pages/integration/config-properties.adoc:
```diff
@@ -28,6 +28,9 @@ The random seed to be used in the solving process.
 implementation during development.
 
+{property_prefix}timefold.solver.{solver_name_prefix}constraint-stream-profiling-enabled::
+Enable xref:constraints-and-score/performance.adoc#constraintProfiling[constraint profiling] to identify the constraints taking the most time in score calculation and thus are worth optimizing.
```
```suggestion
Enable xref:constraints-and-score/performance.adoc#constraintProfiling[constraint profiling] to identify the constraints taking the most time in score calculation and thus might be worth optimizing.
```

Some constraints are just intrinsically harder. It's not because they take more time that they are "always" worth optimizing.

Christopher-Chianelli committed:
feat: add constraint stream profiling

Profiling constraints is notorishing difficult, since each
component of a constraint are converted to nodes, some of which
are shared. As such, a JVM method profile is basically unreadable and
does not represent how much time is actually spent for each constraint.

To aid in profiling, an optional constraintStreamProfilingEnabled
configuration was added. If set to true,
it wraps each tuple lifecycle node inside a ProfilingTupleLifecycle,
which will measure how long each lifecycle executes.
The ProfilingTupleLifecycle find out what constraint is responsible
for creating that lifecycle by getting snapshot of the stack traces
from its constraint stream's creator (when a constraint stream is
shared, their stack traces are merged into the same set).

At the end of solving, a profiling summary is then produced in
the INFO log.

timefold.solver.constraint-stream-profiling-mode was added as
a property to Quarkus and Spring Boot to configure profiling
(defaults to false).

Christopher-Chianelli committed:
fix: remove duplicate anchor

triceo committed:
chore: adapt to the new report

triceo committed:
chore: properly identify input nodes

triceo committed:
chore: adjust the docs

triceo committed:
chore: formatting

Christopher-Chianelli committed:
docs: update docs to match new format

Christopher-Chianelli reviewed (commented):
Changes LGTM (Cannot approve as I am the author of the PR)

triceo committed:
Update the examples

triceo committed:
Review comments
